dali/imgcodec/decoders/libtiff/tiff_libtiff.cc

```diff
@@ -304,23 +330,52 @@ DecodeResult LibTiffDecoderInstance::Decode(DecodeContext ctx,
       using InputType = detail::depth2type<InTypeBits>::type;
 
       kernels::DynamicScratchpad scratchpad;
-      InputType *row_in;
+      InputType *in;
       if (info.bit_depth == InTypeBits) {
-        row_in = static_cast<InputType *>(row_buf.get());
+        in = static_cast<InputType *>(buf.get());
       } else {
-        row_in = static_cast<InputType*>(scratchpad.Alloc(mm::memory_kind_id::host,
-                   volume(in_line_shape) * sizeof(InputType), sizeof(InputType)));
+        in = static_cast<InputType*>(scratchpad.Alloc(mm::memory_kind_id::host,
+                  volume(tile_shape) * sizeof(InputType), sizeof(InputType)));
       }
 
       OutType *const img_out = out.mutable_data<OutType>();
-      for (int64_t roi_y = 0; roi_y < roi.shape()[0]; roi_y++) {
-        LIBTIFF_CALL(TIFFReadScanline(tiff.get(), row_buf.get(), roi.begin[0] + roi_y, 0));
-        if (info.bit_depth != InTypeBits) {
-          detail::UnpackBits(info.bit_depth, row_in, row_buf.get(), volume(in_line_shape));
+
+      int64_t first_tile_y = roi.begin[0] - roi.begin[0] % info.tile_height;
+      int64_t first_tile_x = roi.begin[1] - roi.begin[1] % info.tile_width;
+
+      for (int64_t tile_y = first_tile_y; tile_y < roi.end[0]; tile_y += info.tile_height) {
+        for (int64_t tile_x = first_tile_x; tile_x < roi.end[1]; tile_x += info.tile_width) {
+          TensorShape<> tile_begin = {tile_y, tile_x};
+          TensorShape<> tile_end = {tile_y + info.tile_height, tile_x + info.tile_width};
+
+          tile_begin[0] = std::max(tile_begin[0], roi.begin[0]);
+          tile_begin[1] = std::max(tile_begin[1], roi.begin[1]);
+          tile_end[0] = std::min(tile_end[0], roi.end[0]);
+          tile_end[1] = std::min(tile_end[1], roi.end[1]);
+
+          TensorShape<> out_shape = {tile_end[0] - tile_begin[0], tile_end[1] - tile_begin[1],
+                                     out_channels};
+
+          if (info.is_tiled) {
+            DALI_ENFORCE(TIFFReadTile(tiff.get(), buf.get(), tile_x, tile_y, 0, 0) > 0,
+                         "TIFFReadTile failed");
+          } else {
+            LIBTIFF_CALL(TIFFReadScanline(tiff.get(), buf.get(), tile_y, 0));
```

Comment:
This looks weird, that we use both LIBTIFF_CALL and DALI_ENFORCE for TIFFRead***

Reply:
That *is* weird, but that's libtiff. Almost all of its functions, including `TIFFReadScanline`, [return 1 on success](http://www.libtiff.org/man/TIFFReadScanline.3t.html) and that's what `LIBTIFF_CALL` checks for. But `TIFFReadTile` [returns the number of bytes read](http://www.libtiff.org/man/TIFFReadScanline.3t.html), because why not